sync-todo/v2/client/cpp/managers/database_manager.cpp

```diff
@@ -0,0 +1,148 @@
+void DatabaseManager::remove(realm::managed<realm::Item> itemToDelete) {
+  _database->write([&]{
+    _database->remove(itemToDelete);
+  });
+}
+
+/** Mark an item as "Completed" when the user presses "C" on a selected item. */
+void DatabaseManager::markComplete(realm::managed<realm::Item> itemToMarkComplete) {
+  _database->write([&]{
+    itemToMarkComplete.isComplete = !itemToMarkComplete.isComplete;
+  });
+}
+
+/** Refresh the database from the UI runloop to show data that has synced in the background. */
+void DatabaseManager::refreshDatabase() {
+  _database->refresh();
+}
+
+/** Toggling offline mode simulates having no network connection by pausing sync.
+ *  The user can write to the database on device, and the data syncs automatically when sync is resumed. */
+void DatabaseManager::toggleOfflineMode() {
+  auto syncSession = _database->get_sync_session();
+  if (syncSession->state() == realm::internal::bridge::sync_session::state::paused) {
+    syncSession->resume();
+    _homeControllerState->offlineModeSelection = OfflineModeSelection::offlineModeDisabled;
+    _homeControllerState->offlineModeLabel = "Go Offline";
+  } else if (syncSession->state() == realm::internal::bridge::sync_session::state::active) {
+    syncSession->pause();
+    _homeControllerState->offlineModeSelection = OfflineModeSelection::offlineModeEnabled;
+    _homeControllerState->offlineModeLabel = "Go Online";
+  }
+}
+
+/** Changing the database subscriptions changes which data syncs to the device. */
+void DatabaseManager::toggleSubscriptions() {
+  auto syncSession = _database->get_sync_session();
+  if (syncSession->state() == realm::internal::bridge::sync_session::state::paused) {
+    _appState->errorManager->setError("Please go online before changing subscriptions.");
+    return;
+  }
+  // Note the subscription state at the start of the toggle operation.
+  // We'll change it after updating the subscriptions.
+  auto currentSubscriptionState = _homeControllerState->subscriptionSelection;
+
+  _database->subscriptions().update([&](realm::mutable_sync_subscription_set& subs) {
+    // If the currentSubscriptionState is `allItems`, toggling it should show only my items.
+    // Remove the `allItems` subscription and make sure the subscription for the user's items is present.
+    if (currentSubscriptionState == SubscriptionSelection::allItems) {
+      subs.remove(_allItemSubscriptionName);
+      // If there isn't yet a subscription for my own items, add it
+      if (!subs.find(_myItemSubscriptionName)) {
+        subs.add<realm::Item>(_myItemSubscriptionName,
+                              [&](auto &item){
+                                return item.owner_id == _userId;
+                              });
+      }
+      // Update the subscription selection to reflect the new subscription.
+      _homeControllerState->subscriptionSelection = SubscriptionSelection::myItems;
+      _homeControllerState->subscriptionSelectionLabel = "Switch to All";
```

Comment:
...and then here you can add hooks for your delegate like `onSubscriptionSelectionChanged(SubscriptionSelection::myItems);`